### DISCUSSION TITLE
Customizing a physical keyboard

### DISCUSSION THREAD

**Discussion by orielyada** [AnySoftKeyboard/AnySoftKeyboard#3601] [Q&A]
Hi @menny and community,

I'm a HW guy, not a SW dev... never used Android studio and new on Git ¯\_(ツ)_/¯
I'm familiar with Anysoftkeyboard physical KB support since my Samsung Captivate-glide back in ... 2012 !

I'm trying to customize the Unihertz Titan physical keyboard to Hebrew. https://www.unihertz.com/products/titan.
The Titan phone has a very minimal 29-key KB and the common letter 'Tav' is missing as it is normally on the comma key to the right of 'M' and the Titan phone has no dedicated comma key...
I found the Hebrew physical KB mapping xml [here](https://github.com/AnySoftKeyboard/AnySoftKeyboard/blob/main/addons/languages/hebrew/pack/src/main/res/xml/heb_physical.xml )

I've read [CONTRIBUTING.md](https://github.com/AnySoftKeyboard/AnySoftKeyboard/blob/main/addons/CONTRIBUTING.md) and Q#3125 and I'm scared to go through all that as it's all Klingon to me. 
Can anybody help me fork, compile and publish after I figure out a good .xml mapping ?

**Comment by orielyada** [comment]
Update: So I didn't make progress on this issue but the phone in question actually belongs to my good friend, an attorney that phone-types a LOT, and he is willing to sponsor the project for this feature. 
We know which physical keys are mapped to each character based on the blackberry layout and since some basic keys such as punctuation, as well as one hebrew end-character ' ץ ' , don't fit the physical KB, those will have to be in a touch-bar form, also defined.
Please advise

**Reply by menny** [reply]
Sorry for the very late response, I did not notice this.

It is very hard to create a physical mapping layout without having the physical device - we need the scan codes the device is sending.
So, to begin:
- right off the bat, do all the keys on the physical keyboard work as expected? Do they produce the correct Hebrew character?
- if not, which characters are missing?
- which physical keys are not mapped correctly?
- which physical keys are not mapped at all?

**Comment by orielyada** [comment]
Thank you so much Menny!
It's not THAT difficult because I'm not COMPLETELY clueless...

Background and considerations:
The Unihertz Titan has a minimalistic 29-key HW keyboard with 3 rows of 10
keys (spacebar is merged)
This keyboard layout is similar to the Blackberry Passport but thankfully
has additional 5 control keys on a top row including Shift and Alt and the
android keys - switch, home and back.
Each HW key has the Alt-key etched on it in super-script.
The problem is that Hebrew has 22 letters but also 5 end-letters ("Sofiot"
written differently when at the end of a word) that normally have separate
keys. That's 27 keys in total - one more than in english - which messes
with key allocation in ALL keyboards.
In a full size KB, the letter "Tav" ( 'ת' equivalent to 't' and very
common) is allocated to the qwerty comma key, next to the 'M' ... but the
Titan phone has 'Enter' to the right of 'M' and no comma key.
Based on the Blackberry layout, we put Tav a row above, move the 'ך' to 'W'
(as is currently) but also move 'ף' to 'Q' which is more common than 'ץ'
and have no choice but to have 'ץ' in a 1-row touchkey as well as
punctuation marks (notably semicolon is completely missing even from the
alt-key combinations).

So...
We suggest 1-row touchkeys that are always-on, during english and hebrew,
with soft keys for extended touch keyboards, punctuation, lanuage switching
and our neglected 'ץ'.
Additionally, since hebrew is right-to-left, the parentheses are often
backwards in many phones (shalom)  to  )שלום(  which is annoying and should
be reversed.

The linked excel below includes the full set of keycodes from the Titan phone (using KeyEvent Display app)
The english and hebrew allocations using hebrew AnySoftKeyboard and the
changes required for the hebrew allocation for the Titan phone including
their character codes (source https://help.keyman.com/keyboard/sil_hebrew/1.9/sil_hebrew)
Additionally, there is a definition for the 1-row touch-keyboard in english
and hebrew with its character codes too.

[https://docs.google.com/spreadsheets/d/1l-xnuWWEa8r96SBTwqkn1p-ORwLmv55k/edit?usp=share_link&ouid=117328778838683271736&rtpof=true&sd=true](url)

I will help in any way possible.
Toda Raba, Ori and Guy
